lib/web/fetch/decompress.js

```diff
@@ -0,0 +1,95 @@
+const { pipeline } = require('node:stream')
+const zlib = require('node:zlib')
+const { redirectStatusSet, nullBodyStatus } = require('./constants')
+const { createInflate } = require('./util')
+
+/**
+ * Decompress the given response body.
+ * @param {Request} request
+ * @param {Response} response
+ * @returns {ReadableStream<Uint8Array> | null}
+ */
+function decompress (request, response) {
+  if (response.body === null) {
+    return response.body
+  }
+
+  const willFollow =
+    response.headers.get('location') &&
+    request.redirect === 'follow' &&
+    redirectStatusSet.has(response.status)
+
+  if (
+    request.method === 'HEAD' ||
+    request.method === 'CONNECT' ||
+    nullBodyStatus.includes(response.status) ||
+    willFollow
+  ) {
+    return response.body
+  }
+
+  return decompressStream(
+    response.body,
+    response.headers.get('content-encoding')
+  )
+}
+
+/**
+ * Decompress the given stream based on the "Content-Encoding" response header.
+ * @param {ReadableStream<Uint8Array>} stream
+ * @param {string | null} contentEncoding The value of the "Content-Encoding" response header.
+ */
+function decompressStream (stream, contentEncoding) {
+  if (!contentEncoding) {
+    return stream
+  }
+
+  // https://www.rfc-editor.org/rfc/rfc7231#section-3.1.2.1
+  // "All content-coding values are case-insensitive..."
+  const codings = contentEncoding
+    .toLowerCase()
+    .split(',')
+    .map((coding) => coding.trim())
+
+  if (codings.length === 0) {
+    return stream
+  }
+
+  const decoders = []
+
+  for (let i = codings.length - 1; i >= 0; --i) {
+    const coding = codings[i]
+
+    // https://www.rfc-editor.org/rfc/rfc9112.html#section-7.2
+    if (coding === 'x-gzip' || coding === 'gzip') {
+      decoders.push(
+        zlib.createGunzip({
+          // Be less strict when decoding compressed responses, since sometimes
+          // servers send slightly invalid responses that are still accepted
+          // by common browsers.
+          // Always using Z_SYNC_FLUSH is what cURL does.
+          flush: zlib.constants.Z_SYNC_FLUSH,
+          finishFlush: zlib.constants.Z_SYNC_FLUSH
+        })
+      )
+    } else if (coding === 'deflate') {
+      decoders.push(createInflate())
+    } else if (coding === 'br') {
+      decoders.push(zlib.createBrotliDecompress())
+    } else {
+      decoders.length = 0
+      break
```

Comment:
```suggestion
      return null
```

Maybe return null or otherwise there is no way to know if it "failed".

Reply:
Also `decompressStream` should probably not live under `/web/`

Reply:
@ronag, can it fail though? The intention was that, if there's no known compression provided, the stream is returned as-is. 

You can still handle failures as you normally would:

```js
decompressStream(myStream).on('error', callback)
```

Reply:
> if there's no known compression provided

The logic seems to be that if an unknown coding is sent, the body isn't decompressed at all. I'm not sure if that sounds right either.

`Content-Encoding: gzip, zstd, br` -> ignored (zstd is not supported)
`Content-Encoding: gzip, br` -> decompressed

Reply:
As a consumer, I expect the body stream to be returned as-is if:

1. It has no `Content-Encoding` set (no `codings` provided as an argument);
2. It has unknown `Content-Encoding` set or an unknown `coding(s)` is provided as an argument. 

I wouldn't expect it to throw because that's a disruptive behavior that implies I need to add a guard _before_ trying to call `decompress`/`decompressStream`. At least, not at a level this low. 

If the API is potentially disruptive, I need to have an extra check in place to prevent scenarios where decompression is impossible. This puts extra work on me as a consumer but also results in a more frail API since the list of supported encodings is baked in Undici and may change across releases. 

I strongly suggest to return the body stream as-is in those two cases I described above. `decompressString()` must never error, save for the streams errors itself (and those are irrelevant in the context of this function).

Reply:
We have to outline whether we are after a runtime behavior or a DX improvement. I shared my opinion on this in the context of runtime behaviors. 

If we are talking about a DX improvements, as in not silently ignore unknown compression algorithms, allowing bugs to slip in, then printing a warning may be enough. 

Alternatively, `decompressString()` may trigger error on the given stream if an unknown encoding is encountered. I still think that's too disruptive. Would love to have a more detailed, use case-driven discussion.

Reply:
What about making the behaviour optional?

I see the benefit of throwing an error if the decompression is not possible, at the same time to return the decompressed stream and move forward.
Maybe using a flag whereas the default behaviour is to not throw, and throw otherwise (or the other way around) can provide the enough ergonomics to enable these use cases?

Reply:
Would erroring the _returned_ stream be a semantically correct thing to do instead of throwing an error? 

That way, the `decompressStream` function doesn't have to become more complex. It can always error the returned stream if the decompression is impossible. The consumer can then decide how to handle those errors in the `error` listener.

I emphasize the returned stream because the input stream must remain intact so it can be operated with by other agent regardless of the decompression results. I also propose this on the assumption that `pipeline()` returns a _new_ stream, which I believe it does. We can error _that_.

```js
function decompressStream(stream, codings) {
  // ...
  const result = pipeline(...)

  if (noDecompressionPossible) {
    result.destroy(new Error('Unknown compression algorithm')
  }

  return result
}
```

One problem with this approach is that it doesn't fail fast enough. We know if one of the `codings` is not supported when iterating over them, even before the pipeline is created. 

I see two options to solve this:

1. Error the pipeline but it's not fail-fast.
2. Accept additional argument but it makes the function more complex:

```js
function decompressStream(stream, codings, throwOnUnknownCoding) {}
```

Reply:
I can also argue that the function can fail _always_, and the consumer can always make it safe with this:

```js
import { decompress } from 'undici'

function decompressOrReturnAsIs(request, response) {
  try {
    let r = decompress(request, response)
    return r
  } catch { 
    return response.body
  }
}
```

If you think throwing will benefit most consumers, I'm not against it.

Reply:
@metcoder95 @KhafraDev do you agree with my reasoning in the post above?

Reply:
Hah, so there's an existing expectation that decompression is _skipped_ if encountered unknown encoding:

```
 ✖ should skip decompression if unsupported codings
```

I suppose that answers the question.

Reply:
throwing an error will make fetch slower

Reply:
Regarding the decoders customization discussion, I don't think it justifies the cost of making `decompressStream` more complex. The only thing you need is the list of decodings, which you can get using a one-liner:

```js
 const codings = contentEncoding
    .toLowerCase()
    .split(',')
    .map((coding) => coding.trim())
```

The purpose of `decompressStream` is to grab the _exact_ behavior Undici has under the hood. If you need a different behavior, you should (1) parse the `Content-Encoding` header; (2) map encodings to decompression streams with custom options. 

I can see how you'd want to _extend_ or override certain things from the default decompression, and for that we, perhaps, can consider exporting those options separately? 

```js
module.exports = {
  gzipDecompressionOptions: {
    flush: zlib.constants.Z_SYNC_FLUSH,
    finishFlush: zlib.constants.Z_SYNC_FLUSH
  }
}
```

Reply:
Regarding erroring or throwing; As you suggest @kettanaito, I'd like to fail faster and inform the implementer the encoding is not supported without even disturbing the input stream, so fallbacks can be applied if they want to.

For `fetch`, as @KhafraDev points out, the `throw`ing might be slower, so we can disable by default, and not throw at all, but rather ignore the non-supported encodings.

The issue I'd like to avoid, is putting the implementer in the spot of all or nothing.

>Even then, that's all the options Undici uses right now. To me this looks like a nice thing to have without substantial use case argumentation. I lean toward iterating on this when we have more of the latter.

Agree, let's do that 👍